libkbfs/folder_block_ops.go

```diff
@@ -2947,6 +2949,149 @@ func (fbo *folderBlockOps) UpdatePointers(lState *lockState, op op) {
 	}
 }
 
+func (fbo *folderBlockOps) unlinkDuringFastForwardLocked(ctx context.Context,
+	lState *lockState, ref blockRef) {
+	fbo.blockLock.AssertLocked(lState)
+	oldNode := fbo.nodeCache.Get(ref)
+	if oldNode == nil {
+		return
+	}
+	oldPath := fbo.nodeCache.PathFromNode(oldNode)
+	fbo.log.CDebugf(ctx, "Unlinking missing node %s/%v during "+
+		"fast-forward", oldPath, ref)
+	fbo.nodeCache.Unlink(ref, oldPath)
+}
+
+func (fbo *folderBlockOps) fastForwardDirAndChildrenLocked(ctx context.Context,
+	lState *lockState, currDir path, children map[string]map[pathNode]bool,
+	kmd KeyMetadata) ([]NodeChange, error) {
+	fbo.blockLock.AssertLocked(lState)
+	dirBlock, err := fbo.getDirLocked(ctx, lState, kmd, currDir, blockRead)
+	if err != nil {
+		return nil, err
+	}
+
+	prefix := currDir.String()
+
+	// TODO: parallelize me?
+	var changes []NodeChange
+	for child := range children[prefix] {
+		entry, ok := dirBlock.Children[child.Name]
+		if !ok {
+			fbo.unlinkDuringFastForwardLocked(
+				ctx, lState, child.BlockPointer.ref())
+			continue
+		}
+
+		fbo.log.CDebugf(ctx, "Fast-forwarding %v -> %v",
+			child.BlockPointer, entry.BlockPointer)
+		fbo.nodeCache.UpdatePointer(child.BlockPointer.ref(),
+			entry.BlockPointer)
+		node := fbo.nodeCache.Get(entry.BlockPointer.ref())
+		newPath := fbo.nodeCache.PathFromNode(node)
+		if entry.Type == Dir {
+			if node != nil {
+				change := NodeChange{Node: node}
+				for subchild := range children[newPath.String()] {
+					change.DirUpdated = append(change.DirUpdated, subchild.Name)
+				}
+				changes = append(changes, change)
+			}
+
+			childChanges, err := fbo.fastForwardDirAndChildrenLocked(
+				ctx, lState, newPath, children, kmd)
+			if err != nil {
+				return nil, err
+			}
+			changes = append(changes, childChanges...)
+		} else if node != nil {
+			// File.
+			changes = append(changes, NodeChange{
+				Node:        node,
+				FileUpdated: []WriteRange{{Len: 0, Off: 0}},
```

Comment:
Just checking, this write range value is a special one that invalidates the whole file, right? Might be worth a comment reminding of that.